specs/iota-rs-ENGINEERING-SPEC-0000.md

```diff
@@ -143,28 +142,29 @@ Endpoint collection all about GET messages.
 
 ### Parameters
 
-| Field | Required | Type | Definition |
+| Parameter | Required | Type | Definition |
 | - | - | - | - |
 | **message_id** | ✔ | [MessageId] | The identifier of message. |
+| **index** | ✔ | str | An indexation key. |
 
 ### Returns
 
 Depend on the final calling method, users could get different results they need:
 
-- `metadata()`: Return metadata of the message.
-- `data()`: Return a [Message] object.
-- `raw()`: Return the raw data of given message.
-- `children()`: Return the list of [messageId]s that reference a message by its identifier.
+- `metadata(&MessageId)`: Return [MessageMetadata](#MessageMetadata) of the message.
+- `data(&MessageId)`: Return a [Message] object.
+- `raw(&MessageId)`: Return the raw data of given message.
+- `children(&MessageId)`: Return the list of [MessageId]s that reference a message by its identifier.
+- `index(&str)` : Return the list of [MessageId]s that have this str as indexation key
 
 ## `find_messages()`
 
-Find all messages by provided message IDs. This method will try to query multiple nodes if the request amount exceed individual node limit.
+Find all messages by provided message IDs.
 
 ### Parameters
 
-| Field | Required | Type | Definition |
+| Parameter | Required | Type | Definition |
 | - | - | - | - |
-| **indexation_key** | ✘ | [String] | The index key of the indexation payload. |
```

Comment:
Why did we remove this? There's a plethora of use-cases where you want to look stuff up by a indexation key; Why would you have a indexation in the first place if you can't even look up a message with it.

Reply:
Previously we already implemented this w/ indexation_key, as you can see [here](https://github.com/iotaledger/iota.rs/blob/16bb49e781fa322bbf809672cb7a3b9627b6f5e5/iota-client/src/client.rs#L611).
Just it might be a little bit redundant because the user use get_message() first by using the indices to get the MessageIds, then use this APIs to find all the messages by the MessageIds. How do you think? We can add this back.

Reply:
Ah I see, that implementation is a bit weird though, I'd say get_message would return a single message but it's kind of dynamic depending on the method you call after it. Also both message_id and index are required there where it should be one of both probably. There's a bit of overlap between get_message and find_messages; It doesn't feel very natural. I'd probably vouch for merging both into just a find_messages and maybe a get_message that just retrieves a single message with the message id alone. See the Google doc which was the base for this spec for how it was originally envisioned, this reflects this: https://docs.google.com/document/d/1tnVisc4gGSsznrszQzbD5pbZuGYd5i4A2tXlmsKrfmo/edit#heading=h.1tdfysevvv42

Reply:
Thank you for the detailed explanation. I added this back and updated the spec.